x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -666,6 +732,45 @@ boolean tryRead(ByteBuffer buf, long offset) throws IOException {
             return true;
         }
 
+        /**
+         * Populates a range in cache
+         * <p>
+         * If the range to write is already available (or pending to be available) in cache, then the listener is completed immediately. If
+         * the range is not fully available then the listener is completed once the missing gaps are written in cache.
```

Comment:
Why do we invoke the listener immediately when some other thread is already downloading data? You probably think of something in the consumer but I failed to figure out why.

Reply:
I introduced this method to only start the fetching of a range ("populate only") and not fetch, write then read a range ("populate and read"). It does not have to read data back, so it can return immediately if another thread will fetch that range. Note that if opening the shard require to read in the range, it will wait anyway too.

Reply:
That means that if we have two requests happening "at the same time", one listener will be invoked after the download, another will be invoked immediately. I am not sure I understand why?

I wonder if we should make `maybeFetchRegion` fire and forget completely instead? Waiting for the download of the full region may be counter productive to recovery since it may be able to start with just the first few kbs of the region.

Reply:
We discussed this and one option is to make the listener have a Boolean argument indicating whether the download has been done or not and explaining in javadoc that false means something else is downloading the region.

Reply:
Henning and discussed this via another channel. We'll keep the current listener, maybe complete it with a boolean indicating if the range to write has been fetched by the current thread or will be fetched some time in the future by a competing thread. I'll try to improve doc too.

Reply:
I pushed 54f582b5a018ee6f7161c85e0f6fad8bf3dbefc9